Use vectors of RStrings for default driver lists in arch_default.h

sukibaby opened the pull request:
Currently we are storing the default driver lists as a macro, and splitting them into RString vectors every time they are needed. This commit changes the macros in `arch_default.h` into vectors of RStrings, so that they do not need to be splitted each time.

I have updated all references where the default driver lists are being called so that the vector is used directly.

A new function is added to RageUtil to make a compatible vector of RStrings from user input which may be separated with a comma.

RageSoundDriver was refactored to check user input in Preferences.ini against the default driver list, and in case of incorrect spelling or failure, provides a detailed error message (including a list of valid options) in logs so the user can resolve the problem.

`fix_bogus_sound_driver_pref` is no longer needed, since we are now telling the user what their valid options are if the user provides an incorrect entry for the `SoundDrivers` preference, instead of silently attempting to guess what the user wanted. Since I think it's rare that someone wants to specify a driver to begin with, we should let them know their exact options if they put something wrong here, instead of silently failing or using an unwanted driver.

I tested this on Windows by specifying `DirectSound-sw` in Preferences.ini, and it worked as expected. I also tried putting an unusable value in, and got the error in logs letting me know what my valid driver options were. Of course, input and movies all work as expected too.

teejusb commented on src/RageUtil.cpp:
```diff
@@ -843,6 +843,30 @@ void split( const std::wstring &sSource, const std::wstring &sDelimitor, std::ve
 		do_split( sSource, sDelimitor, asAddIt, bIgnoreEmpty );
 }
 
+std::vector<RString> splitToVector( const RString &sSource, const char delimiter, const bool bIgnoreEmpty )
```
Why don't we just use the function above instead of rewriting a new one? If you wanted to write a new one to return a vector then just do:

```cpp
std::vector<RString> splitToVector( const RString &sSource, const char delimiter, const bool bIgnoreEmpty ) {
  std::vector<RString> result;
  split(sSource, delimiter, &result, bIgnoreEmpty);
  return result;
}
```

Writing a brand new implementation is unnecessary.

teejusb commented on src/arch/Sound/RageSoundDriver.cpp:
```diff
@@ -9,6 +9,17 @@
 #include <cstddef>
 #include <vector>
 
+void WarnUserAboutBadSoundDriverEntry()
```
Use anonymous namespace.

teejusb commented on src/arch/arch_default.h:
```diff
@@ -7,20 +7,18 @@
-#define DEFAULT_MOVIE_DRIVER_LIST "FFMpeg,Null"
-#define DEFAULT_SOUND_DRIVER_LIST "WaveOut,DirectSound-sw,WDMKS,Null"
-
+const std::vector<RString> DEFAULT_INPUT_DRIVER_LIST = {"DirectInput", "Pump", "Para"};
```
Vectors and Strings are not trivially destructible.

If you want to do this then you have to do one of two things:
- you can use an array of char* instead
- create a function that returns a static pointer reference
```cpp
std::vector<RString>& DefaultInputDriverList()
{
  static std::vector<RString>* ret = {"DirectInput", "Pump", "Para"};
  return *ret;
}
```
https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

teejusb commented:
Also slight nit to not call the commit `Vectorize` because that's a common machine learning term which my brain goes to LOL. Maybe just say `use vectors` or something

sukibaby commented:
This new commit should resolve the issue of trivial destruction.

teejusb commented on src/arch/arch_default.h:
```diff
@@ -33,21 +57,29 @@
+}
 #endif
-#define DEFAULT_MOVIE_DRIVER_LIST "FFMpeg,Null"
-// PulseAudio is the preferred Unix driver since it allows the gives non
```
The comments are still useful.

sukibaby replied:
my mistake

teejusb commented on src/RageUtil.cpp:
```diff
@@ -843,6 +843,14 @@ void split( const std::wstring &sSource, const std::wstring &sDelimitor, std::ve
 		do_split( sSource, sDelimitor, asAddIt, bIgnoreEmpty );
 }
 
+std::vector<RString> VectorFromString( const RString& sSource, const char delimiter, const bool bIgnoreEmpty )
```
I think let's just call it `split` like the other functions. Honestly ideally they should be named with upper camel case but that's legacy so can be handled it in a separate PR.

teejusb commented on src/arch/InputHandler/InputHandler.cpp:
```diff
@@ -175,9 +175,8 @@ DriverList InputHandler::m_pDriverList;
-	std::vector<RString> DriversToTry;
-	split( drivers, ",", DriversToTry, true );
+	const std::vector<RString>& drivers = drivers_.empty() ? GetDefaultInputDriverList() : VectorFromString(drivers_, ',', true);
+	std::vector<RString> DriversToTry(drivers.begin(), drivers.end());
```
Do we need to create a copy here actually? We can just use `drivers` directly no?

sukibaby replied:
True, this now doesn't make the copy and renames drivers -> driversToTry to meet with the Google style guide.

teejusb commented on src/arch/Sound/RageSoundDriver.cpp:
```diff
@@ -17,34 +31,40 @@ RageSoundDriver *RageSoundDriver::Create( const RString& drivers )
 	std::vector<RString> drivers_to_try;
```
It's kinda weird how `drivers_to_try` is weirdly cased throughout these files. Let's make them the same with proper naming style.

teejusb commented on src/arch/arch_default.h:
```diff
@@ -57,22 +89,22 @@
  * distribute, and/or sell copies of the Software, and to permit persons to
  * whom the Software is furnished to do so, provided that the above
  * copyright notice(s) and this permission notice appear in all copies of
- * the Software and that both the above copyright notice(s) and this
```
For now let's keep these line breaks the same

sukibaby replied:
my mistake

sukibaby committed:
Make static vectors of default driver lists

Currently we are storing the default driver lists as a macro, and splitting them into RString vectors every time they are needed. This commit changes the macros in `arch_default.h` into vectors of RStrings, so that they do not need to be split each time.

I have updated all references where the default driver lists are being called so that the vector is used directly.

A new function is added to RageUtil to make a compatible vector of RStrings from user input which may be separated with a comma.

RageSoundDriver was refactored to check user input in Preferences.ini against the default driver list, and in case of incorrect spelling or failure, provides a detailed error message (including a list of valid options) in logs so the user can resolve the problem.

`fix_bogus_sound_driver_pref` is no longer needed, since we are now telling the user what their valid options are if the user provides an incorrect entry for the `SoundDrivers` preference, instead of silently attempting to guess what the user wanted. Since I think it's rare that someone wants to specify a driver to begin with, we should let them know their exact options if they put something wrong here, instead of silently failing or using an unwanted driver.

I tested this on Windows by specifying `DirectSound-sw` in Preferences.ini, and it worked as expected. I also tried putting an unusable value in, and got the error in logs letting me know what my valid driver options were. Of course, input and movies all work as expected too.
